packages/plugin-react/src/index.ts

```diff
@@ -162,10 +172,16 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
         const isProjectFile =
           !isNodeModules && (id[0] === '\0' || id.startsWith(projectRoot + '/'))
 
+        let babelOptions = staticBabelOptions
         if (typeof opts.babel === 'function') {
-          const resolvedOpts = opts.babel(id, options)
-          babelOptions = createBabelOptions(resolvedOpts)
-          runPluginOverrides(resolvedConfig, babelOptions)
+          const rawOptions = opts.babel(id, { ssr })
+          babelOptions = createBabelOptions(id, rawOptions, ssr)
+          runPluginOverrides(babelOptions)
+        } else if (!babelOptions) {
+          babelOptions = createBabelOptions(id, opts.babel, ssr)
+          if (!runPluginOverrides(babelOptions)) {
+            staticBabelOptions = babelOptions
```

Comment:
I've added an optimization where the `createBabelOptions(opts.babel)` object is reused if no `api.reactBabel` hooks exist and `opts.babel` is *not* a function